controllers/helmchart_controller.go

```diff
@@ -435,19 +423,122 @@ func (r *HelmChartReconciler) reconcileFromTarballArtifact(ctx context.Context,
-		} else if isDir || changed {
-			// Package the chart
-			pkgPath, err = chartutil.Save(helmChart, tmpDir)
+		}
+	}
+
+	isDir := chartFileInfo.IsDir()
+	switch {
+	case isDir:
+		// Load dependencies
+		if err = chartutil.ProcessDependencies(helmChart, helmChart.Values); err != nil {
+			err = fmt.Errorf("failed to process chart dependencies: %w", err)
+			return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+		}
+
+		// Determine chart dependencies
+		deps := helmChart.Dependencies()
+		reqs := helmChart.Metadata.Dependencies
+		lock := helmChart.Lock
+		if lock != nil {
+			// Load from lockfile if exists
+			reqs = lock.Dependencies
+		}
+		var dwr []*DependencyWithRepository
+		for _, dep := range reqs {
+			// Exclude existing dependencies
+			for _, existing := range deps {
+				if existing.Name() == dep.Name {
+					continue
+				}
+			}
+
+			// Continue loop if file scheme detected
+			if strings.HasPrefix(dep.Repository, "file://") {
+				dwr = append(dwr, &DependencyWithRepository{
+					Dependency: dep,
+					Repo:       nil,
+				})
+				continue
+			}
+
+			// Discover existing HelmRepository by URL
+			repository, err := r.resolveDependencyRepository(ctx, dep, chart.Namespace)
+			if err != nil {
+				repository = &sourcev1.HelmRepository{
+					Spec: sourcev1.HelmRepositorySpec{
+						URL: dep.Repository,
+					},
+				}
+			}
+
+			// Configure ChartRepository getter options
+			clientOpts, err := r.getRepositoryClientOptions(ctx, repository)
+			if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.AuthenticationFailedReason, err.Error()), err
+			}
+
+			// Initialize the chart repository and load the index file
+			chartRepo, err := helm.NewChartRepository(repository.Spec.URL, r.Getters, clientOpts)
+			if err != nil {
+				switch err.(type) {
+				case *url.Error:
+					return sourcev1.HelmChartNotReady(chart, sourcev1.URLInvalidReason, err.Error()), err
+				default:
+					return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+				}
+			}
+			if repository.Status.Artifact != nil {
+				indexFile, err := os.Open(r.Storage.LocalPath(*repository.GetArtifact()))
+				if err != nil {
+					return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+				}
+				b, err := ioutil.ReadAll(indexFile)
+				if err != nil {
+					return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+				}
+				if err = chartRepo.LoadIndex(b); err != nil {
+					return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+				}
+			} else {
+				// Download index
+				err = chartRepo.DownloadIndex()
+				if err != nil {
+					return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+				}
+			}
+
+			dwr = append(dwr, &DependencyWithRepository{
+				Dependency: dep,
+				Repo:       chartRepo,
+			})
+		}
```

Comment:
I think this is a candidate to be factored out, but while taking into account the limitations around the `defer` of the client configuration below.

Reply:
The `defer` related limitation kind of makes it hard to extract the whole thing into something meaningful, I've been trying to wrap my brain around it but couldn't yet figure out a good solution.
Let me know what you have in mind.

Reply:
I get what you mean, and can't come up with anything else than collecting the cleanup funcs and `defer`ing that stack. But that's, well... ugly.